### ISSUE TITLE
Conflict due to incompatible Ruby version from dependencies

### ISSUE THREAD

**Issue by ChannyAzar** [bundler/bundler-features#95]
We have a rails application, running on an Amazon instance using Beanstalk service. Recently we ran into a very annoying issue, when deploying our new version to Production.
Our application crashed with the following error:
`Gem::InstallError: net-ssh requires Ruby version >= 2.0.`
We specified exact version of each gem in our Gemfile, but `net-ssh` is a dependency that was upgraded shortly before with new version that required RUBY VERSION: >= 2.0 ...
We can't control the dependencies. Bundler connects to rubygems.org, and find a list of all of the required gems with our specific versions. Because all of the gems in our Gemfile have dependencies of their own (and some of those have their own dependencies), running bundle install on the Gemfile will install quite a few gems. Some of those gems can be upgraded and no longer support our old Ruby version,  or any other gem's version.
It's a situation that we can not predict, as we can not keep track of all of them all the time.

We must avoid this situation. There must be a way...
Can't the bundler manage this? 
Maybe to tell it somehow to install dependencies that only support a certain ruby version? 
We can also in a case of exact version for a gem, to install the dependencies that WAS RELEASED before the date this specified gem's version was released, assuming that if someone specifies a version for a gem, he is aware he should get dependencies around this period. 

Relevant data:
Ruby 1.9.3
Rails 3.2.13
Bundler 1.10.6

Thanks!

**Comment by segiddins** [comment]
This will be handled once the new index is finished -- bundler currently doesn't have the required ruby version available when resolving.

**Comment by segiddins** [comment]
But there's already a PR implementing the required resolver work :)

**Comment by ChannyAzar** [comment]
Good!
Won't it better to implement my last suggestion (to install the dependencies that was released before the date this specified gem's version was released) ?
It seems to me as the expected behavior, if I want a certain version, I would like to get also the dependencies it had previously ...

**Comment by segiddins** [comment]
that's what the lockfile is for! Additionally, that information isn't available at all

**Comment by ChannyAzar** [comment]
That information isn't available at all? or an additional change is needed for it? 
If the released date is available, the exact version in Gemfile can be much more useful ...

Anyway... Is there an expected date to handle this issue?

Thank you!

**Comment by segiddins** [comment]
Whenever the new index work is finished, which is the culmination of a two years worth of effort across the entire bundler team

**Comment by davidvartanian** [comment]
Is there any workaround about this issue?

**Comment by indirect** [comment]
@davidvartanian put a version requirement in your Gemfile, like `gem "foo", "<1.9"`, if you know that 1.9 is the version that starts requiring a newer Ruby than you have.

**Comment by davidvartanian** [comment]
Thanks @indirect !

**Comment by ChannyAzar** [comment]
Note that it will not resolve your problem completely. If the gem `bar` is a dependency of your gem `foo`, and `bar` will start requiring a newer Ruby than you have, specifying `foo` in your Gemfile will not help. You will have to wait for the 'crash' and then, put a version requirement also for `bar`... And what if a dependency of `bar` itself starts requiring a newer Ruby? same thing...! and so on !
Another solution is needed, as I wrote above.

**Comment by shiralwz** [comment]
change the version of net-ssh in Gemfile, 
`group :test, :development do
  gem "rspec-rails", ">= 2.5.0"
  gem 'factory_girl', '2.6.4'
  gem 'factory_girl_rails', '1.7.0'
  gem 'faker', '1.1.2'
  gem 'forgery', '0.5.0'
  gem 'magic_encoding', '0.0.2'
  gem 'email_spec', '1.4.0'

  gem 'net-ssh', '~> 2.0'
  gem 'capistrano', '~> 2.15.7'
  gem 'capistrano-rbenv', '1.0.5'
  gem 'rb-readline','0.4.2'
end`

**Comment by indirect** [comment]
This is theoretically fixed in the 1.12 prerelease. Give it a try with `gem install bundler --pre`.

**Comment by glennpratt** [comment]
This did not seem to work for me, but I didn't track down what the fix was supposed to do:

```
$ gem install bundler --pre
Fetching: bundler-1.12.0.pre.2.gem (100%)
Successfully installed bundler-1.12.0.pre.2
Installing ri documentation for bundler-1.12.0.pre.2
1 gem installed
$ bundle install
Fetching gem metadata from https://rubygems.org/.........
Fetching version metadata from https://rubygems.org/..
Resolving dependencies...
Using rake 10.5.0
Using i18n 0.7.0
Using multi_json 1.11.2
Using builder 3.0.4
Using bundler 1.11.2
Using docile 1.1.5
Using json 1.8.3
Using minitest 5.8.4
Using moped 1.5.3
Using origin 1.1.0
Using tzinfo 0.3.46
Using simplecov-html 0.10.0
Using yard 0.8.7.6
Using activesupport 3.2.22.1
Installing iso8601 0.9.0 (was 0.8.7)

Gem::InstallError: iso8601 requires Ruby version >= 2.0.0.
Installing simplecov 0.11.2
Installing activemodel 3.2.22.1
An error occurred while installing iso8601 (0.9.0), and Bundler cannot continue.
Make sure that `gem install iso8601 -v '0.9.0'` succeeds before bundling.

**Comment by indirect** [comment]
@glennpratt right now, you'll need to add a `ruby "1.9.3"` line to your Gemfile for this to work. We're still discussing how to handle Gemfiles that don't declare a Ruby version.

**Comment by glennpratt** [comment]
Thanks for the quick reply. Like this?

```
$ cat Gemfile.ruby-1.9.rb
source 'https://rubygems.org'

ruby '1.9.3'

# Specify your gem's dependencies in ruby-duration.gemspec
gemspec
$ BUNDLE_GEMFILE=Gemfile.ruby-1.9.rb bundle install
Fetching gem metadata from https://rubygems.org/.........
Fetching version metadata from https://rubygems.org/..
Resolving dependencies...
Using rake 10.5.0
Using i18n 0.7.0
Using multi_json 1.11.2
Using builder 3.0.4
Using bundler 1.11.2
Using docile 1.1.5
Installing iso8601 0.9.0

Gem::InstallError: iso8601 requires Ruby version >= 2.0.0.
Using json 1.8.3
Using minitest 5.8.4
Using moped 1.5.3
Using origin 1.1.0
Using tzinfo 0.3.46
Using simplecov-html 0.10.0
Using yard 0.8.7.6
Using activesupport 3.2.22.2
An error occurred while installing iso8601 (0.9.0), and Bundler cannot continue.
Make sure that `gem install iso8601 -v '0.9.0'` succeeds before bundling.
```

**Comment by indirect** [comment]
@glennpratt yes, except you're using Bundler 1.11, which I already told you won't work. You need the Bundler 1.12 prerelease.

**Comment by glennpratt** [comment]
@indirect oh sorry, this confused me:

```
$ bundler --version
Bundler version 1.12.0.pre.2
$ bundle --version
Bundler version 1.11.2
```

Wow, that was super fast.

```
$ time BUNDLE_GEMFILE=Gemfile.ruby-1.9.rb bundler install
Using rake 10.5.0
Using i18n 0.7.0
Using multi_json 1.11.2
Using builder 3.0.4
Using bundler 1.12.0.pre.2
Using docile 1.1.5
Using iso8601 0.8.7
Using json 1.8.3
Using minitest 5.8.4
Using moped 1.5.3
Using origin 1.1.0
Using tzinfo 0.3.46
Using simplecov-html 0.10.0
Using yard 0.8.7.6
Using activesupport 3.2.22.2
Using ruby-duration 3.2.2 from source at `.`
Using simplecov 0.11.2
Using activemodel 3.2.22.2
Using mongoid 3.1.7
Bundle complete! 7 Gemfile dependencies, 19 gems now installed.
Use `bundle show [gemname]` to see where a bundled gem is installed.

real	0m0.478s
user	0m0.285s
sys	0m0.068s
```

Thanks!

**Comment by indirect** [comment]
@glennpratt glad it worked! I have no idea why your `bundler` and `bundle` are not the same version of Bundler, but that sounds like something has gone terribly wrong with your gem and/or ruby installation. They should always be the same. :)

**Comment by glennpratt** [comment]
I guess I'll blame RVM. I just installed this Ruby via RVM moments ago to test this (I don't use 1.9).
```
$ gem install bundler
Fetching: bundler-1.11.2.gem (100%)
Successfully installed bundler-1.11.2
Installing ri documentation for bundler-1.11.2
1 gem installed
$ bundle --version; bundler --version
Bundler version 1.11.2
Bundler version 1.11.2
$ gem install bundler --pre
Fetching: bundler-1.12.0.pre.2.gem (100%)
Successfully installed bundler-1.12.0.pre.2
Installing ri documentation for bundler-1.12.0.pre.2
1 gem installed
$ bundle --version; bundler --version
Bundler version 1.11.2
Bundler version 1.12.0.pre.2
$ gem uninstall bundler -v 1.11.2
Successfully uninstalled bundler-1.11.2
$ bundle --version; bundler --version
Bundler version 1.12.0.pre.2
Bundler version 1.12.0.pre.2
```

**Comment by mikeyhew** [comment]
This is still unfixed, at least for me with bundler version 1.12.5. The `shopify_api` gem had no ruby dependency in version 4.2.2, but in version 4.3.0 a ruby dependency of >= 2.3 was added. If this is my Gemfile:

``` ruby
source "https://rubygems.org"
gem 'shopify_api'
```

then `bundle install` fails with this error:

```
Gem::InstallError: shopify_api requires Ruby version >= 2.3.
An error occurred while installing shopify_api (4.3.0), and Bundler cannot
continue.
Make sure that `gem install shopify_api -v '4.3.0'` succeeds before bundling.
```

But if I add a version restiction on shopify_api of `< 4.3`, it correctly installs `shopify_api@4.2.2` without errors.

**Comment by segiddins** [comment]
@mikeyhew it's fixed in the 1.13 RC

**Comment by mikeyhew** [comment]
@segiddins glad to hear it! I'll give it a try later and bug you again if there's problems.
